syntax = "proto3";

package api.services.v1;
option go_package = "github.com/ScienceObjectsDB/go-api/api/services/v1";
option java_multiple_files = true;
option java_package = "com.github.ScienceObjectsDB.java_api.api.services.v1";
option java_outer_classname = "DatasetObjectsServices";

import "api/services/v1/dataset_object_service_models.proto";

import "google/api/annotations.proto";

service DatasetObjectsService {
    //CreateObjectGroup Creates a new object group
    rpc CreateObjectGroup(CreateObjectGroupRequest) returns (CreateObjectGroupResponse) {
        option (google.api.http) = {
            post: "/api/v1/objectgroup/create"
            body: "*"
        };
    }

    //CreateObjectGroupVersion Creates a new object group version
    rpc AddRevisionToObjectGroup(AddRevisionToObjectGroupRequest) returns (AddRevisionToObjectGroupResponse) {
        option (google.api.http) = {
            post: "/api/v1/objectgrouprevisions/create"
            body: "*"
        };
    }

    //GetObjectGroup Returns the object group with the given ID
    rpc GetObjectGroup(GetObjectGroupRequest) returns (GetObjectGroupResponse) {
        option (google.api.http) = {
            post: "/api/v1/objectgroup/get"
            body: "*"
        };
    }

    //GetObjectGroupCurrentVersion Returns the head version in the history of a given object group
    rpc GetCurrentObjectGroupRevision(GetCurrentObjectGroupRevisionRequest) returns (GetCurrentObjectGroupRevisionResponse) {
        option (google.api.http) = {
            post: "/api/v1/currentobjectgroup/get"
            body: "*"
        };
    }

    rpc GetObjectGroupRevision(GetObjectGroupRevisionRequest) returns (GetObjectGroupRevisionResponse) {
        option (google.api.http) = {
            post: "/api/v1/objectgrouprevisions/get"
            body: "*"
        };
    }

    rpc GetObjectGroupRevisions(GetObjectGroupRevisionsRequest) returns (GetObjectGroupRevisionsResponse) {
        option (google.api.http) = {
            get: "/api/v1/objectgrouprevisions/get"
        };
    }

    //FinishObjectUpload Finishes the upload process for an object
    rpc FinishObjectUpload(FinishObjectUploadRequest) returns (FinishObjectUploadResponse) {
        option (google.api.http) = {
            post: "/api/v1/object/finish"
            body: "*"
        };
    }

    rpc DeleteObjectGroup(DeleteObjectGroupRequest) returns (DeleteObjectGroupResponse) {
        option (google.api.http) = {
            delete: "/api/v1/objectgroup/{id}"
        };
    }

    rpc DeleteObjectGroupRevision(DeleteObjectGroupRevisionRequest) returns (DeleteObjectGroupRevisionResponse) {
        option (google.api.http) = {
            delete: "/api/v1/objectgrouprevision/{id}"
        };
    }
}